### PR TITLE
Added the ability to convert the View/Edit Data tab into the Query Tool tab #5908

### PR DESCRIPTION
**Pull request by nikhil-mohite**

### CODE REVIEW

File: web/pgadmin/tools/sqleditor/static/js/components/sections/Query.jsx [hunk 1]
```diff
@@ -508,9 +515,51 @@ export default function Query() {
...
+    queryToolCtx.modal.showModal(gettext('Promote to Query Tool'), (closeModal) =>(
+      <ConfirmPromotionContent
+        closeModal={closeModal}
+        text={'Manually editing the query will cause this View/Edit Data tab to be converted to a Query Tool tab. You will be able to edit the query text freely, but no longer be able to use the toolbar buttons for sorting and filtering data. </br> Do you wish to continue?'}
+        onContinue={(formData)=>{
+          promoteToQueryTool();
+          // Use replaceSelection as user can add single character or select text and enter new values
+          editor.current.replaceSelection(event.key);
+          preferencesStore.setPreference(formData);
+          setIsPromoted((prev)=> {
```
Comment (adityatoshniwal) [inline web/pgadmin/tools/sqleditor/static/js/components/sections/Query.jsx:529]: there is no need to maintain state for ispromoted. You can do the keydown unbind there itself.

File: web/pgadmin/tools/sqleditor/utils/query_tool_preferences.py [hunk 2]
```diff
@@ -181,6 +181,17 @@ def register_query_tool_preferences(self):
         )
     )
 
+    self.view_edit_promotion_warning = self.preference.register(
+        'Editor', 'view_edit_promotion_warning',
+        gettext("Show View/Edit Data Promotion Warning ?"),
```
Comment (adityatoshniwal) [inline web/pgadmin/tools/sqleditor/utils/query_tool_preferences.py:186]: remove the space before question mark.

File: web/pgadmin/tools/sqleditor/utils/query_tool_preferences.py [hunk 3]
```diff
@@ -181,6 +181,17 @@ def register_query_tool_preferences(self):
...
+        'boolean', True,
+        category_label=PREF_LABEL_OPTIONS,
+        help_str=gettext(
+            'If set to True, View/Edit Data window will show promote '
```
Comment (adityatoshniwal) [inline web/pgadmin/tools/sqleditor/utils/query_tool_preferences.py:190]: If set to True, View/Edit Data tool will show promote to Query tool confirm dialog on query edit.

**Comment by adityatoshniwal** [comment]
Press backspace and cancel promotion. The query will apply the backspace. I think it should show the cursor but CM should be readonly.

image

**Review by adityatoshniwal — CHANGES_REQUESTED** [review]

**Comment by adityatoshniwal** [comment]
On backspace and continue - it adds "Backspace"

**Comment by nikhil-mohite** [comment]
> Shouldn't we also update the titles and remove the table name? FYI, all the feature tests are failing for this.

Done

**Comment by nikhil-mohite** [comment]
> On backspace and continue - it adds "Backspace" image

Resolved

**Comment by nikhil-mohite** [comment]
> The pop-up appears for literally any key. I tried with simple CMD+Tab or capslock. Otherwise, looks to be working fine.
  
Will check if we can avoid this.

**Comment by adityatoshniwal** [comment]
> > The pop-up appears for literally any key. I tried with simple CMD+Tab or capslock. Otherwise, looks to be working fine.
> 
> Will check if we can avoid this.

The logic I suggest can be - listen the CodeMirror onchange event and based on that show the pop-up. I guess you'll have to keep it editable.

**Comment by nikhil-mohite** [comment]
> > > The pop-up appears for literally any key. I tried with simple CMD+Tab or capslock. Otherwise, looks to be working fine.
> > 
> > 
> > Will check if we can avoid this.
> 
> The logic I suggest can be - listen the CodeMirror onchange event and based on that show the pop-up. I guess you'll have to keep it editable.

Done

**Pull request merged**